apm-agent-core/src/main/java/co/elastic/apm/agent/report/IntakeV2ReportingEventHandler.java

```diff
@@ -83,6 +84,7 @@
     private int errorCount;
     private long gracePeriodEnd;
     private boolean shutDown;
+    private volatile boolean closed;
```

Comment:
Maybe reuse `shutDown` and make that volatile? Closed seems to be used for the same condition.

Reply:
I started with that but then changed it for two reasons:

1. `shutDown` is queried on each event, so making it volatile will have more impact
2. there will be a strange race condition due to https://github.com/elastic/apm-agent-java/blob/64a6c21d3b3cfe50b1fd58e7d8bbe62daccd13b7/apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerReporter.java#L218 

Looking back at it, maybe both reasons are not valid:
1. we actually want to avoid handling events after close, so maybe this query is a good one
2. in this case, this race condition is not a problem as we wait on the disruptor for 5 seconds, but it may be if are not fully aware to that (e.g. if you switch https://github.com/elastic/apm-agent-java/blob/64a6c21d3b3cfe50b1fd58e7d8bbe62daccd13b7/apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerReporter.java#L225 and https://github.com/elastic/apm-agent-java/blob/64a6c21d3b3cfe50b1fd58e7d8bbe62daccd13b7/apm-agent-core/src/main/java/co/elastic/apm/agent/report/ApmServerReporter.java#L221 in order)

what do you think?

Reply:
I think it should wait for the shutdown event to be handled. This ensures that events which are currently in the ring buffer are handled, if possible. If it can't (for example, because the event handler waits because of a grace period), the `shutDown` variable should be set to `true` by the thread handling the shutdown. Maybe we should wait on a future for the shutdown event to be handled, then close the event handler and after that shut down the disruptor.

Reply:
If you want the thread initiating the shutdown to change this flag, then consolidating to one flag is good and no need to wait on a future as we already wait on the disruptor. In the current state- this race condition is a good one, just as you described - either thread can change it but we make sure it is changed by the end of the close method. 
And maybe it is also a good thing that we stop handling event after close is called (even if we pay the price of frequent queries for a volatile).

So, bottom line - I you say we should change it to be one flag that can be changed by close, right?